themes/default/content/learn/abstraction-encapsulation/component-resources/index.md

```diff
@@ -29,18 +29,91 @@ A resource in Pulumi terms is basically a building block. That building block co
 
 Why would you make a component resource instead of just a "simple" logical grouping that you can call and use like a generic library full of classes and functions? A component resource shows up in the Pulumi ecosystem just like any other resource. That means it has a trackable state, appears in diffs, and has a name field you can reference that refers to the entire grouping.
 
-We've actually already started creating a component resource in the encapsulation part when we made a new class that built up an `s3.Bucket` and an `s3.BucketPolicy` with a `define_policy` function. Let's now turn that logical grouping into an actual component resource.
+We've actually already started creating a component resource in the encapsulation part when we made a new class that built up an `s3.Bucket` and an `s3.BucketPolicy`. Let's now turn that logical grouping into an actual component resource.
 
 ## Converting to a component resource
 
 When we're converting to a component resource, we're subclassing the `ComponentResource` so that our new component resource can get all of the lovely benefits of a resource (state tracking, diffs, name fields, etc.) that other resources have.
 
 In Python, we subclass by using `super()` in the initialization of the class. This call ensures that Pulumi registers the component resource as a resource properly.
 
+{{< chooser language "typescript,python" />}}
+
+{{% choosable language typescript %}}
+
+```typescript
+// ...
+
+// Create a class that encapsulates the functionality by subclassing
+// pulumi.ComponentResource.
+class OurBucketComponent extends pulumi.ComponentResource {
+    public bucket: aws.s3.Bucket;
+    private bucketPolicy: aws.s3.BucketPolicy;
+
+    private policies: { [K in PolicyType]: aws.iam.PolicyStatement } = {
+        default: {
+            Effect: "Allow",
+            Principal: "*",
+            Action: [
+                "s3:GetObject"
+            ],
+        },
+        locked: {
+            /* ... */
+        },
+        permissive: {
+            /* ... */
+        },
+    };
+
+    private getBucketPolicy(policyType: PolicyType): aws.iam.PolicyDocument {
+        return {
+            Version: "2012-10-17",
+            Statement: [{
+                ...this.policies[policyType],
+                Resource: [
+                    pulumi.interpolate`${this.bucket.arn}/*`,
+                ],
+            }],
+        }
+    };
+
+    constructor(name: string, args: { policyType: PolicyType }, opts?: pulumi.ComponentResourceOptions) {
```

Comment:
Does TypeScript do exception handling in line, or does it come with the super class? I handled the exceptions in the Python code, so I was wondering if it made sense to handle them here, as well (unless TS doesn't do that).

Reply:
Ah, yes -- if you mean the `KeyError` bit, then yes, we should actually be protected from these kinds of exceptions by the compile-time type-checking we get against the keys in `policies`.

Reply:
Ah, ok. Makes sense. I'll resolve.